### ISSUE TITLE
[Swift][FlightRPC] Processing valid utf-8 lead to misaligned error

### ISSUE THREAD

**Issue by mgrazianoc** [apache/arrow#37884]
### Describe the bug, including details regarding any error messages, version, and platform.

**Bug**

During FlightClient.doGet call, processing utf-8 generated from the flight server using [arrow-rs](https://github.com/apache/arrow-rs) leads to misaligned error.

**Details**

It's a simple `testString` in a single `TEST` column.
The same test was made on the python flight server, and no errors were raised. 

**Error Messages**

```sh
Swift/UnsafeRawPointer.swift:1203: Fatal error: load from misaligned raw pointer
```

**System details**
- Using Xcode Version 15.0 beta 8 (15A5229m)
- Building for Apple Vision Pro Simulator
- Swift-Driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.106 clang-1500.0.40.1)

### Component(s)

Swift

**Comment by TheGuptaEmpire** [comment]
Bug Report:

Summary:
When making a `FlightClient.doGet` call, processing UTF-8 data generated from the flight server using `arrow-rs` library leads to a fatal error due to misaligned raw pointer.

Details:
The issue arises when attempting to process a simple `testString` in a single `TEST` column. This problem specifically occurs in Swift and does not occur when using the Python flight server.

Error Messages:
```
Swift/UnsafeRawPointer.swift:1203: Fatal error: load from misaligned raw pointer
```

System Details:
- Xcode Version: 15.0 beta 8 (15A5229m)
- Building for: Apple Vision Pro Simulator
- Swift-Driver version: 1.87.1
- Apple Swift version: 5.9 (swiftlang-5.9.0.128.106 clang-1500.0.40.1)

Component(s):
- Swift

Steps to Reproduce:

1. Use Xcode Version 15.0 beta 8 (15A5229m) with Swift-Driver version 1.87.1.
2. Set up a FlightClient and attempt a `doGet` call.
3. Ensure the data being processed is UTF-8 generated from the flight server using `arrow-rs`.
4. Verify that the data contains a `testString` in a single `TEST` column.

Expected Results:
The `doGet` call should process the UTF-8 data without encountering any misaligned raw pointer errors.

Actual Results:
A fatal error occurs due to a misaligned raw pointer during the processing of the UTF-8 data.

Additional Notes:
- This issue does not occur when using the Python flight server, indicating that it is specific to the Swift implementation.
- It is important to investigate and resolve the misalignment of raw pointers when processing the UTF-8 data in Swift.

**Comment by lukedg97** [comment]
This issue was almost fixed in [flatbuffers#8061](https://github.com/google/flatbuffers/pull/8061)

However the highlighted variable here:
<img width="1092" alt="image" src="https://github.com/apache/arrow/assets/8648152/b73d9b42-1195-49f8-84cf-e623e65417c9">

is never `True` as far as I can tell in that code. 

When `allowReadingUnalignedBuffers` is replaced with `True` that fixes the problem

**Comment by abandy** [comment]
I believe if we change are readers to set the above flag on the ByteBuffer then that might resolve the issue.  I will update the code once the issues for [linting](https://github.com/apache/arrow/issues/38131) and [Invalid RecordBatch offsets](https://github.com/apache/arrow/issues/37726) have been merged
